doc/ag.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "AG" "1" "September 2013" "" ""
.
.SH "NAME"
\fBag\fR \- The Silver Searcher\. Like ack, but faster\.
.
.SH "SYNOPSIS"
\fBag\fR [\fIoptions\fR] PATTERN [PATH]
.
.SH "DESCRIPTION"
Recursively search for PATTERN in PATH\. Like grep or ack, but faster\.
.
.SH "OPTIONS"
\fB\-\-ackmate\fR:
.
.br
\~\~\~\~ Output results in a format parseable by AckMate \fIhttps://github\.com/protocool/AckMate\fR\.
.
.P
\fB\-a \-\-all\-types\fR:
.
.br
\~\~\~\~ Search all files\. This doesn\'t include hidden files, and also doesn\'t respect any ignore files
.
.P
\fB\-A \-\-after [LINES]\fR:
.
.br
\~\~\~\~ Print lines after match\. Defaults to 2\.
.
.P
\fB\-B \-\-before [LINES]\fR:
.
.br
\~\~\~\~ Print lines before match\. Defaults to 2\.
.
.P
\fB\-\-[no]break\fR:
.
.br
\~\~\~\~ Print a newline between matches in different files\. Enabled by default\.
.
.P
\fB\-\-[no]color\fR:
.
.br
\~\~\~\~ Print color codes in results\. Enabled by default\.
.
.P
\fB\-\-color\-line\-number\fR:
.
.br
\~\~\~\~ Color codes for line numbers\. Defaults to 1;33\.
.
.P
\fB\-\-color\-match\fR:
.
.br
\~\~\~\~ Color codes for result match numbers\. Defaults to 30;43\.
.
.P
\fB\-\-color\-path\fR:
.
.br
\~\~\~\~ Color codes for path names\. Defaults to 1;32\.
.
.P
\fB\-\-column\fR:
.
.br
\~\~\~\~ Print column numbers in results\.
.
.P
\fB\-C \-\-context [LINES]\fR:
.
.br
\~\~\~\~ Print lines before and after matches\. Defaults to 2\.
.
.P
\fB\-D \-\-debug\fR:
.
.br
\~\~\~\~ Output ridiculous amounts of debugging info\. Probably not useful\.
.
.P
\fB\-\-depth NUM\fR:
.
.br
\~\~\~\~ Search up to NUM directories deep\. Default is 25\.
.
.P
\fB\-f \-\-follow\fR:
.
.br
\~\~\~\~ Follow symlinks\.
.
.P
\fB\-F \-\-fuzzy [ERRORS]\fR:
.
.br
\~\~\~\~ Use fuzzy matching\. Match pattern within a given number of errors\.
.
.P
\fB\-\-[no]group\fR
.
.br
\fB\-g PATTERN\fR:
.
.br
\~\~\~\~ Print filenames matching PATTERN\.
.
.P
\fB\-G\fR, \fB\-\-file\-search\-regex PATTERN\fR:
.
.br
\~\~\~\~ Only search filenames matching PATTERN\.
.
.P
\fB\-\-[no]heading\fR
.
.br
\fB\-\-hidden\fR:
.
.br
\~\~\~\~ Search hidden files\. This option obeys ignore files\.
.
.P
\fB\-\-ignore PATTERN\fR:
.
.br
\~\~\~\~ Ignore files/directories matching this pattern\. Literal file and directory names are also allowed\.
.
.P
\fB\-\-ignore\-dir NAME\fR:
.
.br
\~\~\~\~ Alias for \-\-ignore for compatibility with ack\.
.
.P
\fB\-i \-\-ignore\-case\fR:
.
.br
\~\~\~\~ Match case insensitively\.
.
.P
\fB\-l \-\-files\-with\-matches\fR:
.
.br
\~\~\~\~ Only print filenames containing matches, not matching lines\. An empty query will print all files that would be searched\.
.
.P
\fB\-L \-\-files\-without\-matches\fR:
.
.br
\~\~\~\~ Only print filenames that don\'t contain matches\.
.
.P
\fB\-m \-\-max\-count NUM\fR:
.
.br
\~\~\~\~ Skip the rest of a file after NUM matches\. Default is 10,000\.
.
.P
\fB\-\-no\-numbers\fR:
.
.br
\~\~\~\~ Don\'t show line numbers
.
.P
\fB\-p \-\-path\-to\-agignore STRING\fR:
.
.br
\~\~\~\~ Provide a path to a specific \.agignore file\.
.
.P
\fB\-\-pager COMMAND\fR:
.
.br
\~\~\~\~ Use a pager such as less\. Use \fB\-\-nopager\fR to override\. This option is also ignored if output is piped to another program\.
.
.P
\fB\-\-print\-long\-lines\fR:
.
.br
\~\~\~\~ Print matches on very long lines (> 2k characters by default)
.
.P
\fB\-Q \-\-literal\fR:
.
.br
\~\~\~\~ Do not parse PATTERN as a regular expression\. Try to match it literally\.
.
.P
\fB\-s \-\-case\-sensitive\fR:
.
.br
\~\~\~\~ Match case sensitively\. Enabled by default\.
.
.P
\fB\-S \-\-smart\-case\fR:
.
.br
\~\~\~\~ Match case sensitively if there are any uppercase letters in PATTERN, or case insensitively otherwise\.
.
.P
\fB\-\-search\-binary\fR:
.
.br
\~\~\~\~ Search binary files for matches\.
.
.P
\fB\-\-stats\fR:
.
.br
\~\~\~\~ Print stats (files scanned, time taken, etc)
.
.P
\fB\-t \-\-all\-text\fR:
.
.br
\~\~\~\~ Search all text files\. This doesn\'t include hidden files\.
.
.P
\fB\-u \-\-unrestricted\fR:
.
.br
\~\~\~\~ Search \fIall\fR files\. This ignores \.agignore, \.gitignore, etc\. It searches binary and hidden files as well\.
.
.P
\fB\-U \-\-skip\-vcs\-ignores\fR:
.
.br
\~\~\~\~ Ignore VCS ignore files (\.gitigore, \.hgignore, svn:ignore), but still use \.agignore\.
.
.P
\fB\-v \-\-invert\-match\fR
.
.br
\fB\-w \-\-word\-regexp\fR:
.
.br
\~\~\~\~ Only match whole words\.
.
.P
By default, ag will ignore files matched by patterns in \.gitignore, \.hgignore, or \.agignore\. These files can be anywhere in the directories being searched\. Ag also ignores files matched by the svn:ignore property in subversion repositories\. Finally, ag looks in $HOME/\.agignore for ignore patterns\. Binary files are ignored by default as well\.
.
.P
If you want to ignore \.gitignore, \.hgignore, and svn:ignore but still take \.agignore into account, use \fB\-U\fR\.
.
.P
Use the \fB\-t\fR option to search all text files, \fB\-a\fR to search all files, and \fB\-u\fR to search all including hidden files\.
.
.SH "EXAMPLES"
\fBag printf\fR: Find matches for "printf" in the current directory\.
.
.P
\fBag foo /bar/\fR: Find matches for "foo" in path /bar/\.
.
.SH "SEE ALSO"
grep(1)


ag.bashcomp.sh
have ag &&
_ag() {
  local lngopt shtopt split=false
  local cur prev

  COMPREPLY=()
  cur=$(_get_cword "=")
  prev="${COMP_WORDS[COMP_CWORD-1]}"

  _expand || return 0

  lngopt='
    --ackmate
    --all-text
    --all-types
    --after
    --before
    --break
    --nobreak
    --case-sensitive
    --color-line-number
    --color-match
    --color-path
    --color
    --nocolor
    --column
    --context
    --debug
    --depth
    --file-search-regex
    --files-with-matches
    --files-without-matches
    --follow
    --fuzzy
    --group
    --nogroup
    --heading
    --noheading
    --help
    --hidden
    --ignore
    --ignore-case
    --ignore-dir
    --invert-match
    --line-numbers
    --literal
    --max-count
    --no-numbers
    --pager
    --nopager
    --parallel
    --path-to-agignore
    --print-long-lines
    --recurse
    --no-recurse
    --search-binary
    --search-files
    --search-zip
    --silent
    --skip-vcs-ignores
    --smart-case
    --stats
    --unrestricted
    --version
    --word-regexp
    --workers
  '
  shtopt='
    -a -A -B -C -D
    -f -F -g -G -h
    -i -l -L -m -n
    -p -Q -r -R -s
    -S -t -u -U -v
    -V -w -z
  '

  # these options require an argument
  if [[ "${prev}" == -@(A|B|C|G|g|m) ]] ; then
    return 0
  fi

  _split_longopt && split=true

  case "${prev}" in
    --ignore-dir) # directory completion
              _filedir -d
              return 0;;
    --path-to-agignore) # file completion
              _filedir
              return 0;;
    --pager) # command completion
              COMPREPLY=( $(compgen -c -- "${cur}") )
              return 0;;
    --ackmate-dir-filter|--after|--before|--color-*|--context|--depth\
    |--file-search-regex|--ignore|--max-count|--workers)
              return 0;;
  esac

  $split && return 0

  case "${cur}" in
    -*)
          if [[ "${COMP_CWORD}" -eq 1 ]] ; then
            COMPREPLY=( $(compgen -W \
              "${lngopt} ${shtopt}" -- "${cur}") )
          else
            COMPREPLY=( $(compgen -W \
              "${lngopt} ${shtopt}" -- "${cur}") )
          fi
          return 0;;
    *)
          _filedir
          return 0;;
  esac
} &&
complete -F _ag ${nospace} ag
